[Caleb Chan] iP

calebcjl opened the pull request.

calebcjl committed:
Rename chatbot to Nano and add greeting and exit message

calebcjl committed:
Add echo user input feature and exit condition

calebcjl committed:
Change echo feature to store user input in a list

calebcjl committed:
Add task class and feature to mark tasks

calebcjl committed:
Tweak to coding standard and display number of completed tasks in list

calebcjl committed:
Refactor code to improve quality

ngyida commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,42 @@
+        completedTaskCount += 1;
+    }
+
+    public void incompleteTask() {
```
This method name might be confusing as incomplete is not a verb.

ngyida replied:
The naming for this method may be confusing as it is not a verb.

ngyida commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,42 @@
+        return this.name;
+    }
+
+    public boolean getTaskCompletionStatus() {
```
This method name might be confusing as it does not sound like a boolean e.g. isTaskCompleted

ngyida replied:
The naming for this method may be confusing as it does not sound like a boolean.

ngyida commented on src/main/java/Nano.java:
```diff
@@ -0,0 +1,178 @@
+        }
+    }
+
+    private static void unmarkTask(String taskName) {
+        int taskIndex = getTaskIndex(taskName);
+        if (taskIndex == -1) {
+            System.out.println(MESSAGE_TASK_NOT_FOUND);
+        } else {
+            if (tasks[taskIndex].getTaskCompletionStatus()) {
+                tasks[taskIndex].incompleteTask();
+                System.out.println(taskName + MESSAGE_TASK_UNMARKED);
+            } else {
+                System.out.println(taskName + MESSAGE_TASK_ALREADY_DONE);
+            }
+        }
+    }
```
The heavy nesting here can be hard for others to understand the code logic. Apart from reducing the nesting level, perhaps this can be improved using better method naming for the Task class and also replacing the magic number "-1" with a constant variable name.

ngyida commented on src/main/java/Nano.java:
```diff
@@ -0,0 +1,178 @@
+    private static Task[] tasks;
+    public static void main(String[] args) {
+        //chatbot startup
+        displayWelcomeMessage();
+        initialiseTaskList();
+
+        //receive commands
+        while (true) {
+            String userInput = getUserInput();
+            executeCommand(userInput);
```
This is a very neat code that uses SLAP.

ngyida reviewed (commented):
Mostly clean code that is easy to follow. Minor amendments to naming and cosmetics requested.

heejet commented on src/main/java/Nano.java:
```diff
@@ -0,0 +1,178 @@
+            } else {
+                System.out.println(taskName + MESSAGE_TASK_ALREADY_DONE);
+            }
+        }
```
Consider using guard clauses here to make the happy path of the method more visible.

heejet commented on src/main/java/Nano.java:
```diff
@@ -0,0 +1,178 @@
+    public static final String MESSAGE_TASK_NOT_FOUND = "Task not found.";
+    public static final String MESSAGE_TASK_MARKED = "Great job completing ";
+    public static final String MESSAGE_TASK_ALREADY_DONE = " is already done";
+    public static final String MESSAGE_TASK_UNMARKED = " is marked as undone";
```
I like how the constant variables are named in upper case, keep up the good work

heejet commented on src/main/java/Nano.java:
```diff
@@ -0,0 +1,178 @@
+        default:
+            displayHelpMessage();
+            break;
+        }
```
I like how the indentation for this switch statement follows the coding standards

heejet commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,42 @@
+public class Task {
+    private String name;
+    private boolean isCompleted;
```
Good job with naming the members and classes

heejet commented on src/main/java/Nano.java:
```diff
@@ -0,0 +1,178 @@
+import java.util.Scanner;
```
I like how the imported classes are listed explicitly

heejet commented on src/main/java/Nano.java:
```diff
@@ -0,0 +1,178 @@
+        }
+    }
+
+    private static void unmarkTask(String taskName) {
```
I like how the method names are verbs

heejet reviewed (commented):
Overall good job following the java coding standards. Variables names are in camelCase, constant names are in uppercase with underscore to separate the words. Method names are verbs.

calebcjl committed:
Improve code based on peer reviews

calebcjl committed:
Add new classes to represent different types of Task

New classes are:
1. Deadline - Task with due date
2. Event - Task with start and end date
3. Todo - regular Task

calebcjl committed:
Add exceptions to detect incorrect input

Exceptions added:
1.  DukeCommandExeception: user entered invalid command
2.  DukeInputFormatException: user input has wrong format

calebcjl committed:
Organise classes into packages

calebcjl committed:
Add /help command to list out all possible commands

calebcjl committed:
Merge branch 'branch-A-Package'

calebcjl committed:
Merge branch 'branch-Level-5'

calebcjl committed:
Create new package for Nano exceptions

siyuancheng178 reviewed (commented):
Great! Very concise and readable code. No coding violation found

calebcjl committed:
Use ArrayList to store tasks instead of fixed array

calebcjl committed:
Add delete feature

calebcjl committed:
Add feature to save and retrieve user task list data

calebcjl committed:
Merge branch 'branch-Level-6'

calebcjl committed:
Merge branch 'branch-Level-7'

calebcjl committed:
Refactor code to include new classes and packages for more OOP

calebcjl committed:
Add feature to find and list tasks containing a keyword

calebcjl committed:
Merge pull request #1 from calebcjl/branch-Level-9

Add feature to find and list tasks containing a keyword

calebcjl committed:
Merge branch 'master' of https://github.com/calebcjl/ip

calebcjl committed:
Add JavaDoc comments to code

calebcjl committed:
Add user guide for Nano

calebcjl committed:
Update the correct README for user guide
